In [2]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.neighbors import *
from sklearn.metrics import *
from sklearn.linear_model import *
import matplotlib.pyplot as plt

Reading data

In [3]:
df = pd.read_csv("California_Houses.csv")

In [4]:
df.nunique()
df.isnull().sum()

Median_House_Value          0
Median_Income               0
Median_Age                  0
Tot_Rooms                   0
Tot_Bedrooms                0
Population                  0
Households                  0
Latitude                    0
Longitude                   0
Distance_to_coast           0
Distance_to_LA              0
Distance_to_SanDiego        0
Distance_to_SanJose         0
Distance_to_SanFrancisco    0
dtype: int64

Dividing data into features and labels

In [5]:
y=df.iloc[:,0]
X=df.iloc[:,1:14]

Splitting the data

In [6]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3, random_state = 100, shuffle = True)
X_val, X_test, y_val, y_test = train_test_split(X_test, y_test, test_size = 0.5, random_state = 100,shuffle=True)

Normalizing data

In [7]:
scaler=StandardScaler()
X_train=scaler.fit_transform(X_train)
X_test=scaler.transform(X_test)
X_val=scaler.transform(X_val)

<h1>Linear regression</h1>

In [8]:
linear_model=LinearRegression()

In [9]:
linear_model.fit(X_train,y_train)

LinearRegression()

In [ ]:
y_preds_val=linear_model.predict(X_val)
y_preds=linear_model.predict(X_test)

# mse and mae for validation set
mse_val=mean_squared_error(y_preds_val,y_val)
mae_val=mean_squared_error(y_preds_val,y_val)



# mse and mae for test set
mse=mean_squared_error(y_preds,y_test)
mae=mean_squared_error(y_preds,y_test)


In [11]:
print("Validation errors : ")
print(f"Mean Squared Error {mse_val} ")
print(f"Mean Absolute Error {mae_val} ")
print("---------------------------------------")
print("Test Errors : ")
print(f"Mean Squared Error {mse} ")
print(f"Mean Absolute Error {mae} ")

Validation errors : 
Mean Squared Error 4537986820.675176 
Mean Absolute Error 4537986820.675176 
---------------------------------------
Test Errors : 
Mean Squared Error 4377867249.641456 
Mean Absolute Error 4377867249.641456 


<h1>Ridge Regression</h1>

<h2>ridge regression model using alpha=1.0</h2>

In [12]:
ridge = Ridge(alpha=1)
ridge.fit(X_train, y_train)

Ridge(alpha=1)

In [13]:
y_preds_val=ridge.predict(X_val)

In [14]:
mse_val=mean_squared_error(y_val,y_preds_val)
mse_scores={ridge:mse_val}
mae_val=mean_absolute_error(y_val,y_preds_val)
mae_scores={ridge:mae_val}

<h2>ridge regression model using alpha=10</h2>

In [15]:
ridge_1 = Ridge(alpha=10)
ridge_1.fit(X_train, y_train)

Ridge(alpha=10)

In [16]:
y_preds_val=ridge_1.predict(X_val)

In [18]:
mse_val=mean_squared_error(y_val,y_preds_val)
mse_scores[ridge_1]=mse_val
mae_val=mean_absolute_error(y_val,y_preds_val)
mae_scores[ridge_1]=mae_val



<h2>ridge regression model using alpha=100<h2>


In [19]:
ridge_2 = Ridge(alpha=100)
ridge_2.fit(X_train, y_train)

Ridge(alpha=100)

In [20]:
y_preds_val=ridge_2.predict(X_val)

In [21]:
mse_val=mean_squared_error(y_val,y_preds_val)
mse_scores[ridge_2]=mse_val
mae_val=mean_absolute_error(y_val,y_preds_val)
mae_scores[ridge_2]=mae_val



<h2>ridge regression model using alpha=1000</h2
>


In [22]:
ridge_3 = Ridge(alpha=1000)
ridge_3.fit(X_train, y_train)

Ridge(alpha=1000)

In [23]:
y_preds_val=ridge_3.predict(X_val)

In [25]:
mse_val=mean_squared_error(y_val,y_preds_val)
mse_scores[ridge_3]=mse_val
mae_val=mean_absolute_error(y_val,y_preds_val)
mae_scores[ridge_3]=mae_val

In [26]:
mse_scores

{Ridge(alpha=1): 4538962578.863141,
 Ridge(alpha=10): 4547355593.219384,
 Ridge(alpha=100): 4593210524.991863,
 Ridge(alpha=1000): 4812798048.653835}

In [43]:
mae_scores

{Ridge(alpha=1): 49631.155848899594,
 Ridge(alpha=10): 49719.09549080721,
 Ridge(alpha=100): 50116.83094488498,
 Ridge(alpha=1000): 51572.0849742724}

In [33]:
best_fit_model=min(mse_scores, key=mse_scores.get)
y_preds=best_fit_model.predict(X_test)

mse=mean_squared_error(y_preds,y_test)
mae=mean_absolute_error(y_preds,y_test)

best_fit_model

print(f"Test Error-------------")
print(f"Mean Squared Error: {mse}")
print(f"Mean Absolute Error: {mae}")

Test Error-------------
Mean Squared Error: 4377890123.917159
Mean Absolute Error: 49080.720808534425


<h1>Lasso Regression</h1>

In [45]:
lasso=Lasso(alpha=1,max_iter=10000)
lasso.fit(X_train,y_train)

Lasso(alpha=1, max_iter=10000)

In [46]:
y_preds_val=lasso.predict(X_val)
mse_val=mean_squared_error(y_val,y_preds_val)
mae_val=mean_absolute_error(y_val, y_preds_val)
mse_scores={lasso:mse_val}
mae_scores={lasso:mae_val}




<h2>Training the model with alpha = 10</h2>


In [47]:
lasso_2=Lasso(alpha=10,max_iter=10000)
lasso_2.fit(X_train,y_train)

Lasso(alpha=10, max_iter=10000)

In [48]:
y_preds_val=lasso_2.predict(X_val)
mse_val=mean_squared_error(y_val,y_preds_val)
mae_val=mean_absolute_error(y_val, y_preds_val)
mse_scores[lasso_2]=mse_val
mae_scores[lasso_2]=mae_val




<h2>Training the model with alpha = 100</h2>


In [49]:
lasso_3=Lasso(alpha=100,max_iter=10000)
lasso_3.fit(X_train,y_train)

Lasso(alpha=100, max_iter=10000)

In [50]:
y_preds_val=lasso_3.predict(X_val)
mse_val=mean_squared_error(y_val,y_preds_val)
mae_val=mean_absolute_error(y_val, y_preds_val)
mse_scores[lasso_3]=mse_val
mae_scores[lasso_3]=mae_val




<h2>Training the model at alpha = 1000</h2>


In [51]:
lasso_4=Lasso(alpha=1000,max_iter=10000)
lasso_4.fit(X_train,y_train)

Lasso(alpha=1000, max_iter=10000)

In [52]:
y_preds_val=lasso_4.predict(X_val)
mse_val=mean_squared_error(y_val,y_preds_val)
mae_val=mean_absolute_error(y_val, y_preds_val)
mse_scores[lasso_4]=mse_val
mae_scores[lasso_4]=mae_val

In [53]:
mse_scores

{Lasso(alpha=1, max_iter=10000): 4538214299.06032,
 Lasso(alpha=10, max_iter=10000): 4540409928.904955,
 Lasso(alpha=100, max_iter=10000): 4577011132.289106,
 Lasso(alpha=1000, max_iter=10000): 4712409487.435132}

In [54]:
mae_scores

{Lasso(alpha=1, max_iter=10000): 49622.5993134719,
 Lasso(alpha=10, max_iter=10000): 49644.75256908696,
 Lasso(alpha=100, max_iter=10000): 49947.12044000822,
 Lasso(alpha=1000, max_iter=10000): 50826.19470164871}

In [ ]:
best_model = min(mse_scores, key=mse_scores.get)
best_model

Lasso(alpha=1, max_iter=10000)

In [ ]:
y_preds=best_model.predict(X_test)
mse=mean_squared_error(y_preds,y_test)
mae=mean_absolute_error(y_preds,y_test)

In [ ]:
print(f"Test Error-------------")
print(f"Mean Squared Error : {mae}")
print(f"Mean Absolute Error : {mse}")

Test Error-------------
Mean Squared Error : 49077.11293259761
Mean Absolute Error : 4377830761.426957


======================================================================================================================================================================

<h3>MSE AND MAE for linear regression</h3>


Without regularization, it might yield the lowest training MSE and MAE but could have higher test errors if it overfits.

<!-- mse and mae are exactly the same for linear regression in the validatio set and the test set

MSE is Typically Larger than MAE
but this is a special case where all residuals are either zero or eaxctly the same non-zero value -->


in validation set<br>
mse = 4377867249.641456<br>
mae = 4377867249.641456



======================================================================================================================================================================

<h3>MSE AND MAE for ridge regression</h3>

<h4>l2 regularization is used:</h4>
adds penalty proportional to the sqaure of the coefficients

best mse error was at alpha = 1


in validation set<br>
lowest mse = 4538962578.863141<br>
lowest mae = 49631.155848899594


==========================================================================================================================================================================

<h3>MSE AND MAE for lasso regression</h3>

<h4> l1 regularization is used</h4>
adds penalty proportional to the absloute of the coefficients

best mse error was at alpha = 1

in validation set<br>
lowest mse = 4538214299.06032<br>
lowes mae = 49622.5993134719

==========================================================================================================================================================================

for lasso and ridge regressions since alpha is the same therefor amount of regularization is the same for both regressions in this dataset

mse is smaller than mae  which means that the error is too small since mse squares the values